dll/shellext/netshell/enumlist.cpp

```diff
@@ -1,5 +1,163 @@
+    if (!pdata)
+        return NULL;
+    return (PWCHAR)&pidl->mkid.abID[pdata->uDeviceNameOffset];
+}
+
+PITEMID_CHILD ILCreateNetConnectItem(INetConnection * pItem)
+{
+    PITEMID_CHILD pidl;
+    ULONG_PTR size;
+    NETCON_PROPERTIES * pProperties;
+    PNETCONIDSTRUCT pnetid;
+    PWCHAR pwchName;
+
+    if (pItem->GetProperties(&pProperties) != S_OK)
+        return NULL;
+
+    size = sizeof(WORD); /* nr of bytes in this item */
+    size += sizeof(NETCONIDSTRUCT);
+    size += (wcslen(pProperties->pszwName) + 1) * sizeof(WCHAR);
+    size += (wcslen(pProperties->pszwDeviceName) + 1) * sizeof(WCHAR);
+
+    /* Allocate enough memory for the trailing id which will indicate that this is a simple id */
+    pidl = static_cast<LPITEMIDLIST>(SHAlloc(size + sizeof(SHITEMID)));
+    pidl->mkid.cb = (WORD)size;
+    pidl->mkid.abID[0] = 0x99;
+
+    /* Copy the connection properties */
+    pnetid = ILGetConnData(pidl);
+    pnetid->guidId = pProperties->guidId;
+    pnetid->Status = pProperties->Status;
+    pnetid->MediaType = pProperties->MediaType;
+    pnetid->dwCharacter = pProperties->dwCharacter;
+    pnetid->uNameOffset = sizeof(NETCONIDSTRUCT);
+    pnetid->uDeviceNameOffset = pnetid->uNameOffset + (wcslen(pProperties->pszwName) + 1) * sizeof(WCHAR);
+
+    pwchName = ILGetConnName(pidl);
+    wcscpy(pwchName, pProperties->pszwName);
+
+    pwchName = ILGetDeviceName(pidl);
+    wcscpy(pwchName, pProperties->pszwDeviceName);
+
+    /* Set the trailing id to null */
+    memset((void*)((ULONG_PTR)pidl + size), 0, sizeof(SHITEMID));
+
+    NcFreeNetconProperties(pProperties);
+
+    return pidl;
+}
+
+HRESULT ILGetConnection(PCITEMID_CHILD pidl, INetConnection ** pItem)
+{
+    HRESULT hr;
+    CComPtr<INetConnectionManager> pNetConMan;
+    CComPtr<IEnumNetConnection> pEnumCon;
+    CComPtr<INetConnection> INetCon;
+    ULONG Count;
+    NETCON_PROPERTIES * pProperties;
+
+    PNETCONIDSTRUCT pdata = ILGetConnData(pidl);
+    if (!pdata)
+        return E_FAIL;
+
+    /* get an instance to of IConnectionManager */
+    hr = CNetConnectionManager_CreateInstance(IID_PPV_ARG(INetConnectionManager, &pNetConMan));
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    hr = pNetConMan->EnumConnections(NCME_DEFAULT, &pEnumCon);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    while (TRUE)
+    {
+        hr = pEnumCon->Next(1, &INetCon, &Count);
+        if (hr != S_OK)
+            return E_FAIL;
+
+        hr = INetCon->GetProperties(&pProperties);
+        if (FAILED_UNEXPECTEDLY(hr))
+            continue;
+
+        BOOL bSame = !memcmp(&pProperties->guidId, &pdata->guidId, sizeof(GUID));
+
+        NcFreeNetconProperties(pProperties);
+
+        if (bSame)
+        {
+            *pItem = INetCon.Detach();
+            return S_OK;
+        }
+    }
+
+    return E_FAIL;
+}
+
+typedef struct tagENUMLIST
+{
+    struct tagENUMLIST *pNext;
+    PITEMID_CHILD pidl;
+} ENUMLIST, *LPENUMLIST;
```

Comment:
If you give this thing a destructor that cleans up `pNext`, you can just delete your head.